rust/feature-flags/src/flags/flag_matching_utils.rs

```diff
@@ -628,15 +705,43 @@ pub async fn should_write_hash_key_override(
             })?;
             conn_timer.fin();
 
-            let rows = sqlx::query(query)
+            // Step 1: Get person data and existing overrides
+            let person_data_rows = sqlx::query(person_data_query)
                 .bind(team_id)
                 .bind(&distinct_ids)
+                .fetch_all(&mut *conn)
+                .await
+                .map_err(|e| {
+                    FlagError::DatabaseError(format!("Failed to fetch person data: {e}"))
+                })?;
+
+            // If no person_ids found, there's nothing to check
+            if person_data_rows.is_empty() {
+                return Ok(false);
+            }
+
+            // Extract existing flag keys from overrides
+            let existing_flag_keys: HashSet<String> = person_data_rows
+                .iter()
+                .filter_map(|row| row.try_get::<String, _>("feature_flag_key").ok())
+                .collect();
+
+            // Step 2: Get active feature flags with experience continuity
+            let flag_rows = sqlx::query(flags_query)
```

Comment:
yeah this is better, nice work.

Last bit of feedback – can we add some specific timing traces here?  Given that we're changing some data access patterns in a potentially scary way, I'd love to actually see how expensive these operations could be. We might be able to infer it from existing code (I think we time this span already), so this is maybe just a "double-check this" comment.

Otherwise no blockers, ship it!